### DISCUSSION TITLE
Changes, required by using 3.0.2 version

### DISCUSSION THREAD

**Discussion by Pablo2048** [mathieucarbou/ESPAsyncWebServer#48] [General]
Hi guys,
thanks for the hard work. This is no an issue, but info about what need to be done if one want to use latest 3.0.2 version in ESP8266 and ESP32 mixed environment.

  * as mentioned in doc some of parameters are now `const` (ie `AsyncWebParameter * p = request->getParam(i);` is now `const AsyncWebParameter * p = request->getParam(i);`
  * many `_P` methods for ESP32 are missing ie. `request->send_P(...` is now `request->send(...` but only for ESP32. For ESP8266 you have to still use `_P(...`
  * unfortunately to send array from progmem has now to be replaced by `beginResponse` ie `request->send_P(200, TEXTHTML, index_htm, index_htm_len);` is now `request->beginResponse(200, TEXTHTML, index_htm, index_htm_len);` but again only for the ESP32.

That is all I found out so far. Now I'm testing long term stability...

**Comment by vortigont** [comment]
Hi @Pablo2048, are you using same code for both platforms in your project?
I mean for ESP32 you still have those `F("somethings")`?

**Comment by mathieucarbou** [comment]
Hi @Pablo2048 : Infos in the release notes ;-)

https://github.com/mathieucarbou/ESPAsyncWebServer/releases

Don't forget that as per the semver convention each major digit increase is a breaking change, and you can control dependency lookup with for example `^3.0.2`.

FYI there is a pending PR that is pending review from @vortigont before I merge it that will put back the _P methods but with a deprecation notice to help ease the pain for people ;-)

**Comment by mathieucarbou** [comment]
Moving to discussion as this is not an issue report.

**Comment by Pablo2048** [comment]
So far I've changed things like this:
```cpp
#if defined(ESP32)
            request->send(200, "application/json", "{}");
#else
            request->send_P(200, PSTR("application/json"), PSTR("{}"));
#endif
```
and
```cpp
#if defined(ESP32)
        request->beginResponse(200, TEXTHTML, index_htm, index_htm_len);
#else
        request->send_P(200, TEXTHTML, index_htm, index_htm_len);
#endif
```

**Reply by mathieucarbou** [reply]
> So far I've changed things like this:
> 
> ```c++
> #if defined(ESP32)
>             request->send(200, "application/json", "{}");
> #else
>             request->send_P(200, PSTR("application/json"), PSTR("{}"));
> #endif
> ```
> 
> and
> 
> ```c++
> #if defined(ESP32)
>         request->beginResponse(200, TEXTHTML, index_htm, index_htm_len);
> #else
>         request->send_P(200, TEXTHTML, index_htm, index_htm_len);
> #endif
> ```

That's the way to go if you need to support multiple platforms. And it will also decrease your firmware size.

**Comment by vortigont** [comment]
@mathieucarbou any chance you might be interested in this? - [Community Platformio support for Arduino core 3.0.3
](https://github.com/espressif/arduino-esp32/discussions/10039)
we are gathering a team of crazy enthusiasts in discord channel. Feel free to join.

**Reply by proddy** [reply]
cool, I joined. Nice to see all these legendary names in one place.

**Comment by proddy** [comment]
Hi @mathieucarbou @vortigont - as a huge user of ESPAsyncWS I've been following this project closely - I haven't yet made the switch over to this fork (I'll need to do a PR for msgPack support first). 

I was wondering, with the recent refactoring have you done any memory profiling? Firmware size is not a real issue for me as my projects all run on ESP32-S3 with 16MB + extra PSRAM. But available heap memory and the max alloc buffer is an important factor, which is why I didn't go with PsychicHTTP.

**Reply by mathieucarbou** [reply]
> Hi @mathieucarbou @vortigont - as a huge user of ESPAsyncWS I've been following this project closely - I haven't yet made the switch over to this fork (I'll need to do a PR for msgPack support first).
> 
> I was wondering, with the recent refactoring have you done any memory profiling? Firmware size is not a real issue for me as my projects all run on ESP32-S3 with 16MB + extra PSRAM. But available heap memory and the max alloc buffer is an important factor, which is why I didn't go with PsychicHTTP.

I did, but not precisely (I am just graphing the heap) and regarding that nothing has changed much from the original fork, except minor improvements.

FYI I am using msgpack also through ArduinoJson and websocket a lot. 
It helped me decrease the message size by 30% and increase the rate.
I don't think ESPAsyncWS needs a change to include msgpack suport: it can be added using ArduinoJson + the binary API of websocket.

Still using old ESP32 nodeMCU dev kit, so I am more limited in RAM.

PsychicHTTP is too memory hungry yes, saw that in the code.

**Reply by proddy** [reply]
I needed to extend the AsyncJsonResponse in `AsyncJson.h` to support msgpack, to be able to serialize the binary compressed data. I can't see any other way of doing it. see https://github.com/emsesp/EMS-ESP32/blob/a7dcb32931361f4fcbaac404100023d69d73f8d3/lib/ESPAsyncWebServer/src/AsyncJson.h#L45

**Reply by mathieucarbou** [reply]
> I needed to extend the AsyncJsonResponse in `AsyncJson.h` to support msgpack, to be able to serialize the binary compressed data. I can't see any other way of doing it. see https://github.com/emsesp/EMS-ESP32/blob/a7dcb32931361f4fcbaac404100023d69d73f8d3/lib/ESPAsyncWebServer/src/AsyncJson.h#L45

Oh right! This part would definitely need an `AsyncMessagePackResponse` and set the content type to `application/msgpack` !

Really looking forward to your contribution 👍

**Reply by vortigont** [reply]
> I needed to extend the AsyncJsonResponse in `AsyncJson.h`

it does not have to be `AsyncJsonResponse`, you can make a class in you project and name it i.e. `AsyncMsgPackResponse` with same API, webserver will accept it via a base class reference. At least as see it.

**Reply by mathieucarbou** [reply]
@proddy @vortigont : I created a ticket for the feature here: https://github.com/mathieucarbou/ESPAsyncWebServer/issues. Let's continue the discussion there :-)

**Reply by mathieucarbou** [reply]
@vortigont @proddy : I've crafted a message Pack support to verify something I was envisioning: message pack being only officially supported over https, the implementation could only be used in the case of  client <-> server binary transfer, like I do for WS. I think the same can be done with Ajax.

If you request a message pack endpoint (with proper content type) from a browser, through http, you'll see something like:![image](https://github.com/user-attachments/assets/3c5ebc8d-5ac0-4d24-98bb-d38fc2943f09)

I've opened a PR here (https://github.com/mathieucarbou/ESPAsyncWebServer/pull/62), I think this still has value, though, limited.

**Comment by vortigont** [comment]
Had not done any profiling, but actually it's all pretty straight forward. Mem is allocated only for tcp socket buffers and on messages enqueue. So as long it does not leak it all depends on Q size and number of clients connected.
I'm wondering if you guys do any kind of message reassembly on the server side? I.e. when ws client send messages TO server with size that does not fit into one ethernet frame (1.5k)?

**Reply by mathieucarbou** [reply]
I personally don't have the use case of big messages from browser to server requiring a reassembly... 

Only have that use case in a file upload endpoint, so I go through a temporary file in the case of file upload, to not use the heap, but in the case of WS, if the message rate is high or might be concurrently used, I wouldn't use the FS.

**Reply by vortigont** [reply]
Faced this necessity long time ago and it was not that trivial to implement actually with possibility to receive data from multiple clients simultaneously. Maybe it might be a good thing to create an abstract implementation in WebSever not bound to json data, someday... But that's a not a common cases, agreed. Usually server sends large data and receives some control data.

**Reply by mathieucarbou** [reply]
> But that's a not a common cases, agreed. Usually server sends large data and receives some control data.

Exactly my thinking too.

**Reply by proddy** [reply]
same here, for file uploads I write to a temp file. I use EventSource for real-time log messaging and these are all < 300 char so no need for re-assembly. If I did have a use case I'd probably use Stream and Print.
